apps/opencs/view/render/terrainselection.cpp

```diff
@@ -0,0 +1,251 @@
+            const auto east = std::find(mSelection.begin(), mSelection.end(), std::make_pair(x + 1, y));
+            const auto west = std::find(mSelection.begin(), mSelection.end(), std::make_pair(x - 1, y));
+            const int textureSizeToLandSizeModifier = 4; // To-do: Calculate modifier according to landSize and landTextureSize
+
+            // Nudge selection by 1/4th of a texture size, similar how blendmaps are nudged
+            #define NUDGEPERCENTAGE (0.25)
+            const int nudgeOffset = (cellSize / landTextureSize) * NUDGEPERCENTAGE;
+            const int landHeightsNudge = (cellSize / landSize)/64; // Does this work with all land size configurations?
+
+            // calculate global vertex coordinates at selection box corners
+            int x1 = x * textureSizeToLandSizeModifier + landHeightsNudge;
+            int x2 = x * textureSizeToLandSizeModifier + textureSizeToLandSizeModifier + landHeightsNudge;
+            int y1 = y * textureSizeToLandSizeModifier - landHeightsNudge;
+            int y2 = y * textureSizeToLandSizeModifier + textureSizeToLandSizeModifier - landHeightsNudge;
+
+            // Draw straigth edges
+            int landHeightX1Y1 = calculateLandHeight(x1, y1),
+                landHeightX1Y2 = calculateLandHeight(x1, y2),
+                landHeightX2Y1 = calculateLandHeight(x2, y1),
+                landHeightX2Y2 = calculateLandHeight(x2, y2);
+
+            if (north == mSelection.end()) {
+                vertices->push_back(osg::Vec3f(toWorldCoords(x) + nudgeOffset, toWorldCoords(y + 1) - nudgeOffset, landHeightX1Y2+5));
+                vertices->push_back(osg::Vec3f(toWorldCoords(x + 1) + nudgeOffset, toWorldCoords(y + 1) - nudgeOffset, landHeightX2Y2+5));
+            }
+
+            if (south == mSelection.end()) {
+                vertices->push_back(osg::Vec3f(toWorldCoords(x) + nudgeOffset, toWorldCoords(y) - nudgeOffset, landHeightX1Y1+5));
+                vertices->push_back(osg::Vec3f(toWorldCoords(x + 1) + nudgeOffset, toWorldCoords(y) - nudgeOffset, landHeightX2Y1+5));
+            }
+
+            if (east == mSelection.end()) {
+                vertices->push_back(osg::Vec3f(toWorldCoords(x + 1) + nudgeOffset, toWorldCoords(y) - nudgeOffset, landHeightX2Y1+5));
+                vertices->push_back(osg::Vec3f(toWorldCoords(x + 1) + nudgeOffset, toWorldCoords(y + 1) - nudgeOffset, landHeightX2Y2+5));
+            }
+
+            if (west == mSelection.end()) {
+                vertices->push_back(osg::Vec3f(toWorldCoords(x) + nudgeOffset, toWorldCoords(y) - nudgeOffset, landHeightX1Y1+5));
+                vertices->push_back(osg::Vec3f(toWorldCoords(x) + nudgeOffset, toWorldCoords(y + 1) - nudgeOffset, landHeightX1Y2+5));
+            }
+        }
+    }
+
+    newGeometry->setVertexArray(vertices);
+
+    const auto drawArrays = new osg::DrawArrays(osg::PrimitiveSet::LINES);
```

Comment:
Not very familiar with osg but this looks fishy (allocating with new and storing in a raw local pointer). Shouldn't that be the osg ref pointer thing?

Reply:
It should be an `osg::ref_ptr` just for the sake of being sure. In this specific case, OSG will take responsibility for deleting things when they're done with (as the object's reference count won't have been incremented when it was constructed, so it won't be asking not to be deleted when the geometry is deleted and deletes its internal vector of `ref_ptr`s), but there's also a good chance that in another similar-looking situation it won't work like that.

In general, it's good to follow the rule that for every `new` a function contains, it also contains a `delete` unless the new thing is held in some kind of smart pointer.